File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummyNode=ListNode(0, head)
        l=dummyNode
        r=head

        while n>0 and r:
            r=r.next
            n-=1
        
        while r:
            l=l.next
            r=r.next
        l.next=l.next.next #Delete the mth node 

        #dummy node contains a node, the pos before head so we return the next node which is the head
        return dummyNode.next



        

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        vector<ListNode*> arr;
        ListNode* temp = head;//using a temp variable, preserves head for a return

        while(temp){
            arr.push_back(temp);
            temp=temp->next;
        }

        int idx = arr.size()-n;//idx to remove
        
        if(idx==0){ //removes the head pointer
            return head->next;
        }

        //linked the node before n to the node after n
        arr[idx-1]->next = arr[idx]->next;

        return head;
    }
};